Prepare for CakePHP 5.x

kevindecapite opened the pull request:
Preparation PR that removes deprecations for upgrading to CakePHP 5.0.0.

kevindecapite committed:
Fix PHPUnit 9 deprecations.

kevindecapite committed:
Fix CakePHP 4.x deprecations.

- TestListenerTrait is used by Cake\TestSuite\Fixture\FixtureInjector; but
- It's removed in CakePHP 5.0.0 and will no longer throw deprecations.

kevindecapite committed:
vendor/bin/phpunit --migrate-configuration

kevindecapite committed:
Upgrade to new fixture management per CakePHP 4.3.0.

kevindecapite committed:
Refactor match() to switch in JsonApiListenerTest.

kevindecapite committed:
Use actions/cache@v4 in GitHub workflow ci.yml.

https://github.blog/changelog/2024-12-05-notice-of-upcoming-releases-and-breaking-changes-for-github-actions/#actions-cache-v1-v2-and-actions-toolkit-cache-package-closing-down

kevindecapite commented:
**Refs**
- https://github.com/FriendsOfCake/crud-json-api/issues/193

kevindecapite committed:
Make PHP 8.1 lowest-supported version.

ADmad commented:
phpunit constraints need to be updated to `^10.5.5 || ^11.1.3 || ^12.0.9` and codesniffer to `^5.1` to get the CI passing.

kevindecapite commented:
@ADmad Thank you. I can't use phpunit 10+ with Cake 4.6 so I assume I should just do the full Cake 5 upgrade in this PR rather than separating the process into two separate ones?

ADmad commented:
Oh, I assumed this was upgrade to Cake 5 since you dropped PHP 7.4 in CI.
The Cake 4 compatible version needs to keep PHP 7.4 support as the core does.

kevindecapite commented:
Originally I was going to do the full upgrade to Cake 5 in a single PR, then I thought I would separate the work into two separate PRs. Now I think I will return to the original idea and do the full upgrade to Cake 5 here.

kevindecapite committed:
Set composer.json requirements:

- cakephp/cakephp:^5.0.0
- friendsofcake/crud:^7.0.0
- phpunit/phpunit:^10.5.5 || ^11.1.3 || ^12.0.9
- friendsofcake/cakephp-test-utilities:^3.0.0
- friendsofcake/search:^7.0.0
- cakephp/cakephp-codesniffer:^5.1
- dms/phpunit-arraysubset-asserts:^0.5.0

kevindecapite committed:
Change type declarations and method signatures per Cake 5 upgrade.

kevindecapite committed:
ORM: fetchTable() instead of ProxyTrait::_table(); orderBy() instead of order().

kevindecapite committed:
Remove RequestHandler component from test_app controllers.

kevindecapite committed:
Update tests/bootstrap.php per Cake 5 upgrade.

kevindecapite committed:
Update PaginationListener.

kevindecapite committed:
Add missing use statements for Cake functions.

kevindecapite committed:
Add missing routes to test_app routes.

kevindecapite committed:
Change JsonApiListener::_getSingleEntity() return to support entity via query.

kevindecapite committed:
Refactor tests per PHPUnit 10.5.5 upgrade.

kevindecapite committed:
Fix PHPUnit 10 deprecations.

kevindecapite commented on src/Listener/JsonApiListener.php:
```diff
@@ -1078,7 +1079,7 @@ protected function _getSingleEntity(Subject $subject): ?EntityInterface
         }
 
-        return $subject->entity;
+        return $subject->get('query') ? $subject->get('query')->first() : $subject->get('entity');
```
Please double-check this because `Subject::get()` does not implement a `null` pattern; but without the check `JsonApiListenerTest::testGetSingleEntity` fails.

ADmad replied:
The possibility of `entity` not being set can be handled using `return $subject->entity ?? null;`

kevindecapite replied:
I didn't encounter a case where `subject->entity` was `null`, only where `subject->query` is `null`. I was asking if the usage of `$subject->get('query')` as the conditional part of the expression is acceptable?

ADmad replied:
Why would you use the query if the entity is already available?

kevindecapite replied:
Sometimes (I don't know all the cases) the entity must come from the query as it isn't set previously.

ADmad replied:
Still as per the current code `$subject->entity` should be checked first and then use the query if entity is not available.

kevindecapite replied:
Looks like tapping into the query isn't needed after all. I think I had a misunderstanding while working in the tests. I ended up going with `return $subject->entity ?? null;` as you suggested.

kevindecapite commented on tests/TestCase/Listener/JsonApiListenerTest.php:
```diff
@@ -1091,24 +1074,21 @@ public function testCheckRequestData()
+                    case 2:
+                    case 3:
+                        return 'POST';
+                }
```
Can be replaced with `match()` now that this is exclusively for Cake 5.x/PHP 8.1.

ADmad commented on src/Plugin.php:
```diff
@@ -13,7 +13,7 @@ class Plugin extends BasePlugin
+     * @var ?string
      */
-    protected $name = 'CrudJsonApi';
+    protected ?string $name = 'CrudJsonApi';
```
While the class name `Plugin` works for BC, it should be changed to `CrudJsonApiPlugin`.

ADmad commented on tests/TestCase/Error/JsonApiExceptionRendererTest.php:
```diff
@@ -5,31 +5,34 @@
 
 class JsonApiExceptionRendererTest extends TestCase
 {
+    use ModelAwareTrait;
```
This is not needed. `TestCase` already uses `LocatorAwareTrait` so you can use `$this->fetchTable()` to get table instances.

ADmad commented on tests/TestCase/Error/JsonApiExceptionRendererTest.php:
```diff
@@ -102,7 +106,7 @@ public function testRenderWithNonValidationError()
     public function testRenderWithValidationError()
     {
-        $countries = TableRegistry::get('Countries');
+        $countries = $this->fetchModel('Countries');
```
```suggestion
        $countries = $this->fetchTable('Countries');
```

ADmad commented:
The `Static Analysis` job is failing due to some psalm issue. We can just drop `pslam` and keep only `phpstan`.

kevindecapite committed:
Remove psalm from GH workflow.

kevindecapite committed:
Apply codesniffer checks & fixes; manual fixes as needed.

kevindecapite committed:
Fix phpstan issues.

kevindecapite committed:
composer require laravel-json-api/neomerx-json-api:^5.0.2

kevindecapite committed:
Fix type declaration breakages from codesniffer cleanup.

kevindecapite committed:
composer require friendsofcake/crud:^7.2.0

kevindecapite committed:
Use SelectQuery instead if Query in JsonApiListener.

kevindecapite committed:
Re-use InstanceConfigTrait in JsonApiView to make getConfig() public again.

kevindecapite committed:
Add _cake_core_ cache config to tests/bootstrap.php for BC.

kevindecapite committed:
Define many-to-many CountriesLanguages through model in test_app.

kevindecapite committed:
Replace switch with match() in JsonApiListenerTest.

kevindecapite committed:
Rename plugin to CrudJsonApiPlugin.

kevindecapite committed:
Remove ModelAwareTrait from JsonApiExceptionRendererTest.

kevindecapite commented:
I'll need help getting the coding standard and static analysis tests passing because code sniffer wants to add typehints that break inheritance requirements.

Regarding phpstan, it seems it doesn't like the magic `serialize()` and `unserialize()` methods.

kevindecapite committed:
Use fully qualified names where appropriate.

kevindecapite commented on composer.json:
```diff
@@ -29,17 +29,17 @@
-        "laravel-json-api/neomerx-json-api": "^5.0"
+        "php": ">=8.1.0",
+        "cakephp/cakephp": "^5.0.0",
+        "friendsofcake/crud": "^7.2.0",
```
Is this ok? It seems needed due to this change:

https://github.com/FriendsOfCake/crud/pull/754/files#diff-5aa78d26d9d06925cb6d1434c1c8afa3962704f1041ef3ecf1cb25561e66ee43L124

kevindecapite committed:
Make entity return check sensible in JsonApiListener::_getSingleEntity().

ADmad committed:
Fix CS errors

ADmad committed:
Fix phpstan errors

ADmad committed:
Fix CS errors

ADmad committed:
Update CI config

kevindecapite commented:
Can I help with anything else or is this finished now?
